compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5721,6 +5668,90 @@ private BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType c
         }
     }
 
+    void solveSelectTypeAndResolveType(BLangQueryExpr queryExpr, BLangExpression selectExp, BType type,
+                                       BType collectionType, List<BType> selectTypes, List<BType> resolvedTypes,
+                                       SymbolEnv env, AnalyzerData data, boolean isReadonly,
+                                       LinkedHashSet<BType> memberTypes) {
+        BType selectType, resolvedType;
+        type = Types.getReferredType(type);
+        switch (type.tag) {
+            case TypeTags.ARRAY:
+                BType elementType = ((BArrayType) type).eType;
+                if (data.commonAnalyzerData.checkWithinQueryExpr) {
+                    selectType = checkExpr(selectExp, env, elementType, data);
+                    BType queryResultType = new BArrayType(selectType);
+                    memberTypes.add(queryResultType);
+                    memberTypes.addAll(data.commonAnalyzerData.checkedErrorList);
+                    resolvedType = getResolvedType(BUnionType.create(null, memberTypes), type, isReadonly, env);
+                } else {
+                    selectType = checkExpr(selectExp, env, elementType, data);
+                    resolvedType = getResolvedType(new BArrayType(selectType), type, isReadonly, env);
+                }
```

Comment:
```suggestion
                BType elementType = ((BArrayType) type).eType;
                selectType = checkExpr(selectExp, env, elementType, data);
                BType queryResultType = new BArrayType(selectType);
                if (data.commonAnalyzerData.checkWithinQueryExpr) {
                    memberTypes.add(queryResultType);
                    memberTypes.addAll(data.commonAnalyzerData.checkedErrorList);
                    queryResultType = BUnionType.create(null, memberTypes);
                } 
                resolvedType = getResolvedType(queryResultType, type, isReadonly, env);
```

Reply:
May be rest of the logic can be improved as well. please check

Reply:
Added in https://github.com/ballerina-platform/ballerina-lang/pull/36576/commits/eac8c96aa024c308022b55c059e78532f19b30b9.